src/acktest/adoption.py

```diff
@@ -0,0 +1,144 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	 http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from abc import abstractmethod, ABC
+from dataclasses import dataclass, asdict
+from typing import Dict, Tuple
+
+from .k8s import resource as k8s
+from .resources import random_suffix_name
+
+ADOPTED_RESOURCE_GROUP = "services.k8s.aws"
+ADOPTED_RESOURCE_VERSION = "v1alpha1"
+ADOPTED_RESOURCE_PLURAL = "adoptedresources"
+ADOPTED_RESOURCE_KIND = "AdoptedResource"
+
+ADOPTED_CONDITION_NAME = "ACK.Adopted"
+
+@dataclass(frozen=True)
+class AdoptedResourceAWSIdentifier:
+    """ Represents the base AWS identifier spec fields from the adopted resource CRD.
+
+    Additional keys need to be configured as allowlisted as part of the 
+    controller generator.
+    """
+    additionalKeys: Dict[str, str]
+
+@dataclass(frozen=True)
+class AdoptedResourceNameOrIDIdentifier(AdoptedResourceAWSIdentifier):
+    nameOrID: str
+
+@dataclass(frozen=True)
+class AdoptedResourceARNIdentifier(AdoptedResourceAWSIdentifier):
+    arn: str
+
+@dataclass(frozen=True)
+class AdoptedResourceKubernetesIdentifiers:
+    """ Represents the Kubernetes spec fields from the adopted resource CRD.
+
+    This class purposefully does not expose metadata configuration elements, as
+    those are configured by the test fixture.
+    """
+    group: str
+    kind: str
+
+@dataclass(frozen=True)
+class AdoptedResourceSpec:
+    """ Represents the adopted resource CRD spec fields.
+    """
+    aws: AdoptedResourceAWSIdentifier
+    kubernetes: AdoptedResourceKubernetesIdentifiers
+
+class AbstractAdoptionTest(ABC):
```

Comment:
In the docstring for abstract base classes, I like to see example usage of the abc. Tell the readers how you expect your class to be instantiated and called in normal operation.